www_src/pages/section-2/grid.jsx

```diff
@@ -0,0 +1,382 @@
+
+    if (pageX > midpointX) {
+      cameraX = (pageX - midpointX) * -1 * this.slotWidth;
+
+      if (isEvenWidth) {
+        cameraX += (this.slotWidth / 2);
+      }
+    } else if (pageX <= midpointX) {
+      cameraX = (midpointX - pageX) * this.slotWidth;
+
+      if (isEvenWidth) {
+        cameraX -= (this.slotWidth / 2);
+      }
+    }
+
+    if (pageY > midpointY) {
+      cameraY = (pageY - midpointY) * -1 * this.slotHeight;
+
+      if (isEvenHeight) {
+        cameraY += (this.slotHeight / 2);
+      }
+    } else if (pageY <= midpointY) {
+      cameraY = (midpointY - pageY) * this.slotHeight;
+
+      if (isEvenHeight) {
+        cameraY -= (this.slotHeight / 2);
+      }
+    }
+
+    var newDisplayState = {
+      zoom: this.state.containerWidth / (this.slotWidth * pagesWide),
+      previousZoom: this.displayState.zoom || this.state.containerWidth / (this.slotWidth * pagesWide),
+      pagesWide: pagesWide,
+      focusedPageCoords: {x: pageX, y: pageY}
+    }
+
+    if (doCentering) {
+      newDisplayState.previousCameraX = this.displayState.cameraX;
+      newDisplayState.previousCameraY = this.displayState.cameraY;
+      newDisplayState.cameraX = cameraX;
+      newDisplayState.cameraY = cameraY;
+    }
+
+    this.displayState = newDisplayState;
+  },
+  onPageClick: function (event) {
+    // TODO : TEMP - Just rotating through zoom factors
+
+    var zoomFactor = 3.25;
+
+    if (this.displayState.pagesWide) {
+      zoomFactor = this.displayState.pagesWide === 1 ? 3.25 : 1;
+    }
+
+    this.activeTile = {
+      x: event.x,
+      y: event.y
+    };
+
+    this.calculateCameraState(event.x, event.y, zoomFactor, true);
+    this.animateCamera(true);
+  },
+  showOverview: function () {
+    // 6 UP
+
+    if (this.activeTile) {
+      this.calculateCameraState(this.activeTile.x, this.activeTile.y, 6.25, true);
+    } else {
+      this.calculateCameraState(Math.floor(this.tilesPerRow / 2 ), Math.floor(this.tilesPerCol / 2), 6.25, true);
+    }
+
+    this.animateCamera(true);
+  },
+  generateGrid: function (width, height) {
+    var grid = [];
+
+    for (var h = 0; h < height; h++) {
+      grid.push([]);
+
+      for (var w = 0; w < width; w++) {
+        grid[h].push(null);
+      }
+    }
+
+    grid[Math.floor(height / 2)][Math.floor(width / 2)] = 'EMPTY';
+
+    return grid;
+  },
+  getInitialState: function () {
+    // 20 x 20 seems to be max before display glitches
+    var layout = this.generateGrid(3, 3);
+
+    return {
+      layout: layout
+    }
+  },
+  addPageClick: function (event) {
+    var self = this;
+
+    var newLayout = this.state.layout;
+
+    newLayout[event.y][event.x] = 'EMPTY';
+
+    var width = newLayout[0].length;
+    var height = newLayout.length;
+
+    // The new page tile's x and y coords may shift if the grid is expanded
+    var newPageX = event.x;
+    var newPageY = event.y;
+
+    // Detect what grid boundaries the given tile touches
+    function edgeDetect(x, y) {
```

Comment:
multiple function declarations in this call suggests this should be a .js lib on its own instead of being intimately tied to the React component. At the very least, those functions can be declared at the same level as addPageClick, so they don't get redefined every time addPageClick is called.